skopt/tests/test_tree_opt.py

```diff
@@ -0,0 +1,102 @@
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_less
+from sklearn.utils.testing import assert_raise_message
+
+from skopt.benchmarks import bench1
+from skopt.benchmarks import bench2
+from skopt.benchmarks import bench3
+from skopt.benchmarks import branin
+from skopt.benchmarks import hart6
+from skopt.tree_opt import gbrt_minimize
+from skopt.tree_opt import forest_minimize
+
+
+def check_no_iterations(minimizer):
+    assert_raise_message(ValueError, "at least one iteration",
+                         minimizer,
+                         branin, [(-5.0, 10.0), (0.0, 15.0)], maxiter=0,
+                         random_state=1)
+
+    assert_raise_message(ValueError, "at least one starting point",
+                         minimizer,
+                         branin, [(-5.0, 10.0), (0.0, 15.0)], n_start=0,
+                         maxiter=2, random_state=1)
+
+
+def test_no_iterations():
+    yield (check_no_iterations, gbrt_minimize)
+    yield (check_no_iterations, partial(forest_minimize, base_estimator='et'))
+    yield (check_no_iterations, partial(forest_minimize, base_estimator='dt'))
+    yield (check_no_iterations, forest_minimize)
+
+
+def test_one_iteration():
+    result = gbrt_minimize(branin, [(-5.0, 10.0), (0.0, 15.0)],
+                           maxiter=1, random_state=1)
+
+    assert_equal(len(result.models), 0)
+    assert_array_equal(result.x_iters.shape, (1, 2))
+    assert_array_equal(result.func_vals.shape, (1,))
+    assert_array_equal(result.x, result.x_iters[np.argmin(result.func_vals)])
+    assert_almost_equal(result.fun, branin(result.x))
+
+
+def test_seven_iterations():
+    result = gbrt_minimize(branin, [(-5.0, 10.0), (0.0, 15.0)],
+                           n_start=3, maxiter=7, random_state=1)
+
+    assert_equal(len(result.models), 4)
+    assert_array_equal(result.x_iters.shape, (7, 2))
+    assert_array_equal(result.func_vals.shape, (7,))
+    assert_array_equal(result.x, result.x_iters[np.argmin(result.func_vals)])
+    assert_almost_equal(result.fun, branin(result.x))
+
+
+def test_forest_minimize_api():
+    # invalid string value
+    assert_raise_message(ValueError,
+                         "Valid values for the base_estimator parameter",
+                         forest_minimize, lambda x: 0., [],
+                         base_estimator='abc')
+
+    # not a string nor a regressor
+    assert_raise_message(ValueError,
+                         "The base_estimator parameter has to either",
+                         forest_minimize, lambda x: 0., [],
+                         base_estimator=42)
+
+    assert_raise_message(ValueError,
+                         "The base_estimator parameter has to either",
+                         forest_minimize, lambda x: 0., [],
+                         base_estimator=DecisionTreeClassifier())
+
+
+def check_minimize(minimizer, func, y_opt, dimensions, margin, maxiter):
+    success = 0
+    N = 3
```

Comment:
Should we increase N to a higher number? I understand that the drawback is that this increases the test time, but then it is better since it can still reach the minimum 2 out of 3 time by chance.